src/scenic/core/lazy_eval.py

```diff
@@ -13,174 +13,195 @@
+            return getattr(self.evaluateIn(context), op)(*subvalues)
+        return DelayedArgument(props, value, _internal=True)
+    return handler
 for op in allowedOperators:
-	setattr(DelayedArgument, op, makeDelayedOperatorHandler(op))
+    setattr(DelayedArgument, op, makeDelayedOperatorHandler(op))
 
 def makeDelayedFunctionCall(func, args, kwargs):
-	"""Utility function for creating a lazily-evaluated function call."""
-	props = set().union(*(requiredProperties(arg)
-	                      for arg in itertools.chain(args, kwargs.values())))
-	def value(context):
-		subvalues = (valueInContext(arg, context) for arg in args)
-		kwsubvals = { name: valueInContext(arg, context) for name, arg in kwargs.items() }
-		return func(*subvalues, **kwsubvals)
-	return DelayedArgument(props, value, _internal=True)
-
-def valueInContext(value, context):
-	"""Evaluate something in the context of an object being constructed."""
-	if needsLazyEvaluation(value):
-		return value.evaluateIn(context)
-	else:
-		return value
+    """Utility function for creating a lazily-evaluated function call."""
+    props = set().union(*(requiredProperties(arg)
+                          for arg in itertools.chain(args, kwargs.values())))
+    def value(context, modifying):
+        subvalues = (valueInContext(arg, context, modifying) for arg in args)
+        kwsubvals = { name: valueInContext(arg, context, modifying) for name, arg in kwargs.items() }
+        return func(*subvalues, **kwsubvals)
+    return DelayedArgument(props, value, _internal=True)
+
+def valueInContext(value, context, modifying={}):
+    """Evaluate something in the context of an object being constructed."""
+    if not needsLazyEvaluation(value):
+        return value
+    elif isinstance(value, LazilyEvaluable):
+        return value.evaluateIn(context, modifying)
+    elif isinstance(value, dict):
+        return { key: valueInContext(val, context, modifying)
+                 for key, val in value.items() }
+    assert False
 
 def requiredProperties(thing):
-	"""Set of properties needed to evaluate the given value, if any."""
-	if isinstance(thing, LazilyEvaluable):
-		return thing._requiredProperties
-	else:
-		return set()
+    """Set of properties needed to evaluate the given value, if any."""
+    if isinstance(thing, LazilyEvaluable):
+        return set(thing._requiredProperties)
+    elif isinstance(thing, dict):
+        properties = set()
+        for key, val in thing.items():
+            properties.update(requiredProperties(key))
+            properties.update(requiredProperties(val))
+        return properties
+    else:
+        return set()
 
 def needsLazyEvaluation(thing):
-	"""Whether the given value requires lazy evaluation."""
-	return isinstance(thing, DelayedArgument) or requiredProperties(thing)
+    """Whether the given value requires lazy evaluation."""
+    if isinstance(thing, dict):
+        return any([needsLazyEvaluation(val) for val in thing.values()])
```

Comment:
You don't need the square brackets here and below (a generator expression is fine). Also I think `requiredProperties(thing)` is fine -- no need for the more verbose `(len(requiredProperties(thing)) > 0)` (if you prefer the latter let's at least drop the unnecessary outer parentheses).